### PR TITLE
feat: introduce `Fluvio Version Manager`

### PR DESCRIPTION
**Pull request by EstebanBorai**
Introduces the FVM (Fluvio Version Manager) with support for:

- Installing `stable`, `latest` and a specific Fluvio Version
- Fresh Fluvio Installer solution
- Switch between Fluvio Versions

### CODE REVIEW

**Review by sehz — CHANGES_REQUESTED** [review]
first pass.  comments are in these groups:
- move to anyhow, that will get rid of vast defined error
- reduce output, don't need to inform user every actions, use tracing
- move all package related abstraction to hub or something.  it's not responsibility of FVM to provide these abstraction
- FVM should deal with channel abstraction not package 
- need more test

**Comment by sehz** [comment]
Probably better break this into smaller PR.

**Comment by EstebanBorai** [comment]
Thanks for the feedback @sehz!

I think moving into a smaller PR will be a good approach to achieve this.

Will apply improvements as I move into smaller PRs.

**Comment by EstebanBorai** [comment]
This PR will be split into the following parts:

- [ ] Introduce FVM for Hub Util
- [ ] FVM
  - [ ] Introduce `fvm self` commands
  - [ ] Introduce `fvm install` command
  - [ ] Introduce `fvm switch` command
  - [ ] Introduce `fvm current` and `fvm show` command

**Comment by sehz** [comment]
Suggest to break down into these ares
- ChannelManager.  This will be responsible for managing channels
- Download/Updater.    Download hub pkg required from ChannelManager
- Then FVM CLI which should be just wrapper for CLI command (show progress, etc).

File: crates/fvm/src/bin/command/install.rs [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+    }
+
+    ///  Performs the installation of the specified `PackageSet`
+    async fn install_package(&self) -> Result<()> {
+        let install_task = InstallTask::new(
+            self.registry.clone(),
+            DEFAULT_PKGSET.to_string(),
+            self.version.clone(),
+        );
+        let pkgset = install_task.fetch_pkgset().await?;
```
[the remark this answers is not in the document]
Comment (digikata) [reply crates/fvm/src/bin/command/install.rs:86]: here pkgset is what is being stored in the hub.  package is a bit misleading but pkgset is a separate term.

File: crates/fluvio-hub-util/src/fvm.rs [hunk 2]
```diff
@@ -40,17 +128,53 @@ impl Display for RustTarget {
...
+            ARM_UNKNOWN_LINUX_GNUEABIHF => Ok(Self::ArmUnknownLinuxGnueabihf),
+            ARMV7_UNKNOWN_LINUX_GNUEABIHF => Ok(Self::Armv7UnknownLinuxGnueabihf),
+            X86_64_APPLE_DARWIN => Ok(Self::X86_64AppleDarwin),
+            AARCH64_APPLE_DARWIN => Ok(Self::Aarch64AppleDarwin),
+            X86_64_PC_WINDOWS_GNU => Ok(Self::X86_64PcWindowsGnu),
+            _ => Err(Error::RustTripleNotSupported(v.to_string())),
+        }
+    }
+}
+
```
[the remark this answers is not in the document]
Comment (EstebanBorai) [reply crates/fluvio-hub-util/src/fvm.rs:145]: This represents the supported build targets for Fluvio as available in CI, took this list from: https://github.com/infinyon/fluvio/blob/f2c49e126c771d58d24d5f5cb0282a6aaa6b23ca/.github/workflows/ci.yml#L141

**Comment by EstebanBorai** [comment]
> Suggest to break down into these ares
> 
> * ChannelManager.  This will be responsible for managing channels
> * Download/Updater.    Download hub pkg required from ChannelManager
> * Then FVM CLI which should be just wrapper for CLI command (show progress, etc).

First decoupling step lives here: https://github.com/infinyon/fluvio/pull/3566 I am aiming to provide FVM/Hub API HTTP Client responsible of fetching binaries URLs from the Hub.

Also unit tests and new type definitions.

**Pull request closed without merging**